Merge remote-tracking branch 'origin/dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality

KuGa-Do opened the pull request:
В этой таске было сделано следующее :
Добавлен класс LoginFailureHandler , который обрабатывает исключения класса SecurityUserService , и сообщает о различных ошибках при попытке 
аутентификации . идея такая - Основных три вида ошибок : введены неправильно данные <- выводим сообщение 
пользователь удалил свой профиль , но еще 30 дней не истекло <- сообщение ссылкой для восстановления , идет по ссылке восстанавливает 
и когда пользователь удалил профиль и 30 дней уже прошли , <- сообщение что " извините профиль был удален ..." за кулисами удаляем его профиль 
Все операции в основном происходят на стороне security . Классы SecurityUserDetails-> LoginFailureSecurity -> login.html
Добавил метод который , раз в день, ходит по базе, и удаляет кастомеров у которых уже истек срок восстановления профиля .

karuksun committed:
небольшой рефакторинг по пути , добавил пару to do , преимнеовал класс чтоб понятней было зачем он

karuksun committed:
комментарии , убрал одно todo

karuksun committed:
Переименовал контроллер , на Oauth2LoginController путало название всегда внего заходил . . .

karuksun committed:
убрал желтую кнопку log in

karuksun committed:
удалил не нужную страницу регистрации

karuksun committed:
Merge branch 'dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality

karuksun committed:
мелкие правки

karuksun committed:
Создал класс login failure handler, убрал маппинг со страницы логина на кнопку входа .

karuksun committed:
Создал поле у кастомера anchorForDelete, у юзера созадл поле isAccountNonBlocked , поменял геттеры сеттеры
создал лог в классе  LoginFailureHandler.java чтоб отслеживать каой там пришел exception , чтоб настраивать было удобней

karuksun committed:
1)метод isAccountNonLocked класса UserDetails, от котрого наследуется наш класс User,
теперь возвращает поле isAccountNonBlockedStatus , это нужно для последующей реализации
в spring security
2) добавлена проверка поля isAccountNonBlockedStatus в классе CustomerServiceImpl.java , там где это нужно.
3) убрал todo
4) в customerPage.html когда разбирался как она работает(чтоб ничего не поломать) по пути добавлял комменты чтоб не теряться .

karuksun committed:
Изменил название MyUserDetailsService на SecurityUserDetailsService так этот сервис нам служит
только для spring security . Так же изменил реализацию , теперь мы проверяем сначал не заблокирован ли пользователь ,
если нет то пускаем дальше .

karuksun committed:
Разделил два exception'a если один то один редирект
если другой то . . . logic LoginFailureHandler <- here

karuksun committed:
Удалил метод checkEmailForRestore в AllUsersRestController , т.к. эту проверку буду делать на беке
до этого на нем , можно сказать весело все секьюрити .... удалил соответсвующие функции в js
поменял модалку , пока работает все четко, удалениен восстановление . . .

karuksun committed:
Exception выбрасывающий исключение, если профиль был удален , по истечении 30 дней

karuksun committed:
Java doc
поправил логику , не входило на админа ...
убрал метод checkCustomerStatus т.к. раньше вся логика в нем весела на кнопке Sign in , сейчас она реализована в классе SecurityUserDetailsService.java
 в методе isExist убрал удаление кастомера у которого прошло 30 дней на восстановление . сейчас логика в SecurityUserDetailsService

karuksun committed:
Merge remote-tracking branch 'origin/dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

karuksun committed:
Merge branch 'dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality

# Conflicts:
#	basic_module/src/main/java/com/jm/online_store/controller/simple/MainPageController.java
#	basic_module/src/main/resources/templates/customerPage.html
#	basic_module/src/main/resources/templates/login.html

karuksun committed:
Merge remote-tracking branch 'origin/dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

karuksun committed:
Создал в CustomerService новый метод deleteAllBlockedWithThirtyDaysPassed()
Который удаляет, кастомеров с у которых, срок 30 дней, для восстановления  прошел .
-Работа шедулера организовна конечно сложновато , я думал сначала просто сделать метод ,
 и повесить аннотацию @Shedule, но тогада в базе в таблице task_settings не отображалась бы эта таска.
 По-этому пришлось все создавать по примеру.
-Создал DeleteExpiredProfileTask и его имплементацию DeleteExpiredProfileTaskImpl
-добавил таску в DataInitializer

karuksun committed:
java.doc

DenisMalikovKzn commented on basic_module/src/main/java/com/jm/online_store/config/handler/LoginFailureHandler.java:
```diff
@@ -0,0 +1,46 @@
+            response.setStatus(HttpStatus.LOCKED.value());
+        } else if (exceptionMessage.contains("Аккаунт был удален")) {
+            response.sendRedirect("/login?accountExpired=true");
+            response.setStatus(HttpStatus.LOCKED.value());
```
тут тогда не LOCKED а DELETED наверное

KuGa-Do replied:
такого статуса к сожалению нет , я сам его хотел , но по большому счету без разницы какой

DenisMalikovKzn reviewed (commented):
вроде все ок

karuksun committed:
Merge branch 'dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality

karuksun committed:
Поправил замечания Ильи , немного поменял логику , как проверить напишу в карточке .

karuksun committed:
Исправил маленький баг

tediik commented on basic_module/src/main/java/com/jm/online_store/config/handler/LoginFailureHandler.java:
```diff
@@ -0,0 +1,46 @@
+                                        AuthenticationException exception) throws IOException, ServletException {
+        String exceptionMessage = exception.getMessage();
+        log.info(exceptionMessage);
+        if (exceptionMessage.contains("Bad credentials")) {
+            response.sendRedirect("/login?loginBadCredentials=true");
+            response.setStatus(HttpStatus.UNAUTHORIZED.value());
+        } else if (exceptionMessage.contains("Аккаунт заблокирован !!!")) {
+            response.sendRedirect("/login?accountBlocked=true");
+            response.setStatus(HttpStatus.LOCKED.value());
+        } else if (exceptionMessage.contains("Аккаунт был удален")) {
+            response.sendRedirect("/login?accountExpired=true");
+            response.setStatus(HttpStatus.LOCKED.value());
```
честно говоря это решение тоже похоже на какой то костыль, мы должны смотреть, что внутри exeption и при этом делать какие то "странные" редиректы

KuGa-Do replied:
В принципе , я сейчас перечитал тз еще раз и честно говоря думал что ты примерно как-то так и хочешь , я вроде когда делал , на созвонах проговаривал свои идеи реализации , ну и ты их вроде одобрял .  Я смотрел примеры  в таком духе , кидаешь эксепшн, ловишь его в loginFailureHandlere,  потом действие какое-то. По этим примерам мандил чтоб к нам встало . В общем если я че-то не правильно понял наведи тогда на мысль или опиши по подробней . Понимаю что сейчас моя реализация далеко не фонтан , если есть идеи как улучшить - жду предложений

tediik replied:
Почему бы нам не ловить тут конкретные exeption, которые ты бросаешь в UserDetailsService? Это возможно? Я спрашиваю, потому что ты лучше знаешь свою реализацию. Т.е. вместо, того, что делать ветвления на основне сообщений в exeption, почему нельзя отталкиваться от самого класса exeption?
И я так и не понял смысл этих редиректов, они нам зачем?

KuGa-Do replied:
редиректы , для того что б всплывали алерты на страничке логина , разный эспешн для разной таблички. Ну а по сообщению просто , легче определять из-чего именно он выпал , ну выводить соответсвующее сообщение юзеру ,

KuGa-Do replied:
вроде сделал так как ты хочешь

tediik commented on basic_module/src/main/java/com/jm/online_store/config/security/AppConfig.java:
```diff
@@ -4,6 +4,9 @@
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
+
+//todo  какой-то непонтяный класс который нигде не используется ... разобраться
+
```
похоже класс просто декларирует нам бин OAuth2Odnoklassniki

KuGa-Do replied:
Исправил

tediik commented on basic_module/src/main/java/com/jm/online_store/config/security/LoginListener.java:
```diff
@@ -18,14 +18,17 @@
     public void onApplicationEvent(InteractiveAuthenticationSuccessEvent event) {
         Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
-        log.debug("Current authentication is " + authentication);
+        log.debug("Current authentication of user" + authentication.getName() + " is");
```
сделай логирование в общепринятом у нас формате с шаблонами подстановки, без конкатенации

KuGa-Do replied:
Исправил

tediik commented on basic_module/src/main/java/com/jm/online_store/service/impl/CustomerServiceImpl.java:
```diff
@@ -261,4 +226,23 @@ public void deleteByID(Long id) {
         customerRepository.deleteById(id);
     }
 
+    /**
+     * Метод который будет ходить по базе раз в день,
+     * и удалять кастомеров которые удалили свой профиль и у которых срок для восстановления истек.
+     * Время и таска создается в Datainitializer'e
+     */
+    @Override
+    @Scheduled(cron = "0 0 0 * * *")
+    @Transactional
```
стой, я был уверен, что у нас есть специальный класс для заданий по расписанию, второе, настройку времени в таком случае вынеси в aplication.yml

KuGa-Do replied:
Поместил период в aplication.yml , проверил с 10 сек периодом - работает

tediik commented on basic_module/src/main/java/com/jm/online_store/service/impl/CustomerServiceImpl.java:
```diff
@@ -261,4 +226,23 @@ public void deleteByID(Long id) {
+        LocalDateTime timeAfterThirtyDaysFromNow = LocalDateTime.now().minusDays(30);
+        List<Customer> listForDelete = customerRepository.findAllByAnchorForDeleteThirtyDays(timeAfterThirtyDaysFromNow);
+        if (listForDelete.isEmpty()) {
+            log.info("Список профилей для удаления пуст, метод для удаления не будет вызвыан");
```
это не очень важная информация, так что давай с уровнем debug и ошибку в слове там поправь

KuGa-Do replied:
Исправил

karuksun committed:
Если считать сверху вниз то исправил 2,3,4,5 замечания Эдуарда

karuksun committed:
Merge branch 'dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality

karuksun committed:
поправил loger

karuksun committed:
Сделал как советовал , Эдуард на созвоне , прошел дебагом по всем сценариям , глянул какие классы Эксепшенов выбрасываются, потсавил обработку
LoginFailureHandler . Убрал сценарий тгго что пользователь заходит если у него истекло 30 дней, т.к. у нас есть шедулер для удаления таких customer'ov.
Все проверил все работает .

karuksun committed:
Merge branch 'dev' into refactoring/profile_recovery_via_springSecurity_regular_functionality

f4nke commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Акции - баги](https://trello.com/c/brEvQVYp/235-%D0%B0%D0%BA%D1%86%D0%B8%D0%B8-%D0%B1%D0%B0%D0%B3%D0%B8)
